src/compiler/checker.ts

```diff
@@ -5119,7 +5119,7 @@ namespace ts {
         }
 
         function getJSInitializerType(decl: Node, symbol: Symbol, init: Expression | undefined): Type | undefined {
-            if (init && isInJavaScriptFile(init) && isObjectLiteralExpression(init) && init.properties.length === 0) {

```

Comment:
This is only called from `getJSSpecialType`, so the `isInJavaScriptFile` check is redundant.
Maybe type annotations should be checked before calling `getJSSpecialType`?

Reply:
Good catch.

I think the right fix might be to only call `getJSInitializerType` on certain kinds of special property assignments. Let me play with that.

Reply:
Yes, getJSInitializerType doesn't make sense for this-property assignments. I'll change this fix.

Reply:
Actually, after thinking a bit more, I think the original fix is better. Any special assignment should use the type annotation if present and not act as an expando object. (It still doesn’t make sense for this-properties to be expando, though.)